mindsdb/interfaces/datastore/datastore.py

```diff
@@ -252,23 +244,30 @@ def save_datasource(self, name, source_type, source, file_path=None):
                 shutil.rmtree(ds_meta_dir)
                 raise Exception('Each column in datasource must have unique non-empty name')
 
-            datasource_record.creation_info = json.dumps(creation_info)
-            datasource_record.data = json.dumps({
-                'source_type': source_type,
-                'source': source,
-                'row_count': len(df),
-                'columns': [dict(name=x) for x in list(df.keys())]
-            })
+            datasource_record = Datasource(
+                company_id=self.company_id,
+                name=name,
+                creation_info=json.dumps(creation_info),
+                data=json.dumps({
+                    'source_type': source_type,
+                    'source': source,
+                    'row_count': len(df),
+                    'columns': [dict(name=x) for x in list(df.keys())]
+                })
+            )
+            session.add(datasource_record)
+            datasource_record = session.query(Datasource).filter_by(company_id=self.company_id, name=name).first()
 
             self.fs_store.put(name, f'datasource_{self.company_id}_{datasource_record.id}', self.dir)
+            session.commit()
 
         except Exception as e:
+            log.error(f'{e}')
             if os.path.isdir(ds_meta_dir):
                 shutil.rmtree(ds_meta_dir)
-            session.delete(datasource_record)
-            raise e
+            session.rollback()
+            raise
```

Comment:
We could just call the internal `delete_datasource` function here, I think, + log the error.